pallets/polkadex-ido/src/lib.rs

```diff
@@ -633,262 +329,103 @@ pub mod pallet {
-        ///
-        /// * `round_id`: Funding round id
-        /// * `beneficiary`: Account Id of Beneficiary
-        #[pallet::weight((10_000, DispatchClass::Normal))]
-        pub fn withdraw_token(origin: OriginFor<T>, round_id: T::Hash, beneficiary: T::AccountId) -> DispatchResult {
+        pub fn withdraw_investment(origin: OriginFor<T>, round_id: T::Hash) -> DispatchResult {
+            let investor_address: T::AccountId = ensure_signed(origin)?;
+            ensure!(<InfoFundingRound<T>>::contains_key(&round_id.clone()), Error::<T>::FundingRoundDoesNotExist);
             let current_block_no = <frame_system::Pallet<T>>::block_number();
-            let creator: T::AccountId = ensure_signed(origin)?;
-            ensure!(<InfoInvestor<T>>::contains_key(&beneficiary), <Error<T>>::InvestorDoesNotExist);
-            ensure!(!<InfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundNotApproved);
-            ensure!(<WhitelistInfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundDoesNotExist);
-            let funding_round = <WhitelistInfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundDoesNotExist)?;
-            ensure!(creator.eq(&funding_round.creator), <Error<T>>::NotACreater);
-            // Check if there is any left to withdraw
-            let total_tokens_bought_by_investors = if T::OnePDEX::get().saturated_into::<BalanceOf<T>>() >= funding_round.token_a_priceper_token_b {
-                funding_round.token_a_price_per_1e12_token_b().saturating_reciprocal_mul(funding_round.amount)
-            } else {
-                funding_round.amount / funding_round.token_a_price_per_1e12_token_b_balance()   //TODO saturated div
-            };
-            let remaining_token = funding_round.amount.saturating_sub(total_tokens_bought_by_investors);
+            let funding_round = <InfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
             ensure!(current_block_no >= funding_round.close_round_block, Error::<T>::WithdrawalBlocked);
-            ensure!(remaining_token > Zero::zero(), Error::<T>::WithdrawalBlocked);
+            let total_raise: u128 = funding_round.actual_raise.saturated_into();
+            let amount: u128 = funding_round.amount.saturated_into();
+            ensure!(total_raise <= amount/2, Error::<T>::CannotWithdrawForSuccesfulIDO);
             let round_account_id = Self::round_account_id(round_id.clone());
-            //Transfers to remaining token back to creator after round.
-            Self::transfer(funding_round.token_a, &round_account_id, &beneficiary, remaining_token.saturated_into())?;
-            Self::deposit_event(Event::WithdrawToken(round_id, creator));
+            let investment = <InvestorInvestment<T>>::get(&round_id, &investor_address);
+            Self::transfer(funding_round.token_b, &round_account_id, &investor_address, investment)?;
+            let empty_raise: BalanceOf<T> = 0_u128.saturated_into();
```

Comment:
Instead of this remove the investor using take method

Reply:
Updated ✅